File: backend/src/modules/business/inventory/models/sparePart.model.js
import mongoose from 'mongoose';

const { Schema } = mongoose;

const imageSchema = new Schema(
  {
    url: { type: String, required: true },
    publicId: { type: String, default: null },
    fileName: { type: String, default: null },
    mimeType: { type: String, default: null },
    uploadedAt: { type: Date, default: Date.now },
  },
  { _id: false }
);

const sparePartSchema = new Schema(
  {
    companyId: {
      type: Schema.Types.ObjectId,
      ref: 'Company',
      required: true,
      index: true,
    },
    partNumber: {
      type: String,
      required: true,
      trim: true,
      uppercase: true,
      maxlength: 100,
    },
    name: {
      type: String,
      required: true,
      trim: true,
      maxlength: 200,
    },
    category: {
      type: String,
      trim: true,
      default: null,
      maxlength: 120,
      index: true,
    },
    description: {
      type: String,
      trim: true,
      default: null,
      maxlength: 5000,
    },
    unit: {
      type: String,
      required: true,
      trim: true,
      maxlength: 40,
      default: 'pcs',
    },
    currentQuantity: {
      type: Number,
      required: true,
      min: 0,
      default: 0,
    },
    minimumQuantity: {
      type: Number,
      required: true,
      min: 0,
      default: 0,
    },
    image: {
      type: imageSchema,
      default: null,
    },
    barcode: {
      type: String,
      trim: true,
      default: null,
      maxlength: 200,
      index: true,
    },
    isActive: {
      type: Boolean,
      default: true,
      index: true,
    },
    deletedAt: {
      type: Date,
      default: null,
      index: true,
    },
  },
  {
    timestamps: true,
    collection: 'spare_parts',
  }
);

sparePartSchema.index(
  { companyId: 1, partNumber: 1 },
  {
    unique: true,
    partialFilterExpression: { deletedAt: null },
  }
);

sparePartSchema.index({ companyId: 1, name: 1, deletedAt: 1 });
sparePartSchema.index({ companyId: 1, category: 1, deletedAt: 1 });
sparePartSchema.index({ companyId: 1, currentQuantity: 1, minimumQuantity: 1 });

const SparePart = mongoose.model('SparePart', sparePartSchema);

export default SparePart;

File: mobile/lib/features/inventory/data/models/spare_part_model.dart
import 'package:mobile/features/inventory/domain/entities/spare_part.dart';

class SparePartModel extends SparePart {
  const SparePartModel({
    required super.id,
    required super.partNumber,
    required super.name,
    required super.unit,
    required super.currentQuantity,
    required super.minimumQuantity,
    required super.stockStatus,
    super.companyId,
    super.category,
    super.description,
    super.image,
    super.barcode,
    super.isActive,
    super.createdAt,
    super.updatedAt,
  });

  factory SparePartModel.fromJson(Map<String, dynamic> json) {
    final imageJson = json['image'];
    return SparePartModel(
      id: json['id']?.toString() ?? '',
      companyId: json['companyId']?.toString(),
      partNumber: json['partNumber']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      category: json['category']?.toString(),
      description: json['description']?.toString(),
      unit: json['unit']?.toString() ?? 'pcs',
      currentQuantity: (json['currentQuantity'] as num?)?.toDouble() ?? 0,
      minimumQuantity: (json['minimumQuantity'] as num?)?.toDouble() ?? 0,
      stockStatus: StockStatus.fromApi(json['stockStatus']?.toString()),
      image: imageJson is Map<String, dynamic>
          ? SparePartImage(
              url: imageJson['url']?.toString() ?? '',
              publicId: imageJson['publicId']?.toString(),
              fileName: imageJson['fileName']?.toString(),
              mimeType: imageJson['mimeType']?.toString(),
              uploadedAt: _parseDate(imageJson['uploadedAt']),
            )
          : null,
      barcode: json['barcode']?.toString(),
      isActive: json['isActive'] != false,
      createdAt: _parseDate(json['createdAt']),
      updatedAt: _parseDate(json['updatedAt']),
    );
  }

  static DateTime? _parseDate(dynamic value) {
    if (value == null) return null;
    return DateTime.tryParse(value.toString());
  }
}
